File: heaps/leastInterval.js
// #include <iostream>
// #include <vector>
// #include <unordered_map>
// using namespace std;

// // Define a struct to hold character and frequency
// struct CharFrequency {
//     char character;
//     int frequency;

//     // Constructor for easy initialization
//     CharFrequency(char ch, int freq) : character(ch), frequency(freq) {}

//     // Overload '<' operator to define comparison for priority_queue
//     bool operator<(const CharFrequency& other) const {
//         // Higher frequency has higher priority
//         return frequency < other.frequency;
//     }
// };

// // Function to convert vector<char>& tasks into a priority queue of CharFrequency objects
// priority_queue<CharFrequency> getCharFrequencyPriorityQueue(const vector<char>& tasks) {
//     // Create an unordered_map to count frequency of each character
//     unordered_map<char, int> frequencyMap;

//     // Count frequencies of each character in tasks
//     for (char ch : tasks) {
//         frequencyMap[ch]++;
//     }

//     // Create a priority_queue to store CharFrequency objects (max heap by default)
//     priority_queue<CharFrequency> pq;

//     // Add CharFrequency objects to priority_queue
//     for (auto pair : frequencyMap) {
//         CharFrequency cf(pair.first, pair.second);
//         pq.push(cf); // Push CharFrequency into priority_queue
//     }

//     return pq;
// }

// class Solution {
// public:
//     int leastInterval(vector<char>& tasks, int n) {
//         priority_queue<CharFrequency> charFrequenciesPQ = getCharFrequencyPriorityQueue(tasks);
//         int result = 0;
//         int k = 0;
//         vector<CharFrequency> temp;
//         unordered_set<char> charSet;

//         while (!charFrequenciesPQ.empty() && !temp.empty()) {
//             if(charFrequenciesPQ.empty()) {
//                 result++;
//                 k++;
//             } else {
//             CharFrequency cf = charFrequenciesPQ.top();
//             charFrequenciesPQ.pop(); // Remove the top element

//             charSet

//             temp.push(cf)

//             }
//         }

//         return 0;
//     }
// };

File: heaps/leastInterval.py
import heapq as hq
from typing import List

class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        if(n == 0):
            return len(tasks)
      
        temp = [0] * 26
        heap = []
        result = 0
        cooldown = n + 1

        for task in tasks:
            temp[ord(task) - 65] += 1

        for val in temp:
            if (val > 0):
                hq.heappush(heap,-1 * val)

        temp = []

        while (len(heap) > 0 or len(temp) > 0):
            wasInserted = False

            if (len(heap) > 0):
                result += 1
                cooldown -= 1
                val = hq.heappop(heap)
                wasInserted = True
                
                if(val + 1 != 0):
                  temp.append(val + 1)

            if(len(temp) > 0 and cooldown > 0 and wasInserted == False) :
                cooldown -= 1
                result += 1
            
            if (cooldown == 0):
                cooldown = n + 1
                while(len(temp) > 0):
                      hq.heappush(heap,val + 1)
        
        return result